{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Standard primers for the Yeast Pathway Kit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pydna"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The six primers below are necessary and sufficient to carry out assemblies with the YPK system."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "p167,p166,p468,p467,p567,p568  =  pydna.parse('''\n",
    "                                                    >167 pCAPSfw (24-mer)\n",
    "                                                    TCCTGACGGGTAATTTTGATTTGC\n",
    "\n",
    "                                                    >166 pCAPSrv (24-mer) EMPTY\n",
    "                                                    CTGTGAAGTGGCTGAAATTTCGTA\n",
    "\n",
    "                                                    >468 pCAPs_release_fw (25-mer) 79.66\n",
    "                                                    gtcgaggaacgccaggttgcccact\n",
    "\n",
    "                                                    >467 pCAPs_release_re (31-mer) \n",
    "                                                    ATTTAAatcctgatgcgtttgtctgcacaga\n",
    "\n",
    "                                                    >567 pCAPsAjiIF (23-mer)\n",
    "                                                    GTcggctgcaggtcactagtgag\n",
    "\n",
    "                                                    >568 pCAPsAjiIR (22-mer)\n",
    "                                                    GTGCcatctgtgcagacaaacg''')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Primer positions in pCAPs derived vectors\n",
    "\n",
    "\n",
    "                                 >-gene-->\n",
    "                >-TP-->           \\     /           >-TP-->\n",
    "                 \\   /             \\   /             \\   /\n",
    "                  \\ /               \\ /               \\ /\n",
    "      167> 511>    |<776             |             777>|    <512     <166\n",
    "     577>      775>|468>         <567|568>         <467|<778     <578    <342\n",
    "                   |                 |                 | \n",
    "     ------------N-Z-----------------A-----------------E-A--------------------\n",
    "    |            o r                 j                 c c                    |\n",
    "    |            t a                 i                 o c                    |\n",
    "    |            I I                 I                 R I                    |\n",
    "    |                                                  V I                    |\n",
    "    |                                                    I                    |\n",
    "    |                                                                         |\n",
    "     --------------------- pYPKa / pYPK0 ------------------------------------\n",
    "    (*) AjiI is only unique in pYPKa and pLBL2, not in pYPK0 and derivatives.\n",
    "    \n",
    "    This is a list of primers for the pYPKa pYPK0 backbone that have been tested\n",
    "    \n",
    "    577 gttctgatcctcgagcatcttaagaattc\n",
    "    167 TCCTGACGGGTAATTTTGATTTGC\n",
    "    511 gtcagcggccgcattgcacagactctgct\n",
    "    775 gcggccgctgacTTAAAT\n",
    "    776 cgttcctcgacTAATTAA\n",
    "    468 gtcgaggaacgccaggttgcccact\n",
    "    567 GTcggctgcaggtcactagtgag\n",
    "    568 GTGCcatctgtgcagacaaacg\n",
    "    467 ATTTAAatcctgatgcgtttgtctgcacaga\n",
    "    777 acgcatcaggatTTAAAT\n",
    "    778 ggtaaatccggatTAATTAA\n",
    "    512 ttcgccaattgattcaggtaaatccggat\n",
    "    578 gttcttgtctcattgccacattcataagt\n",
    "    166 CTGTGAAGTGGCTGAAATTTCGTA\n",
    "    342 CCTTTTTACGGTTCCTGGCCT\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "\n",
    "### Background\n",
    "\n",
    "The yeast pathway kit is a system for the in vivo assembly entire metabolic or\n",
    "other multi gene pathways in Saccharomyces cerevisiae. It has three principal\n",
    "components which are combinations of DNA fragments cloned in the pCAPs positive\n",
    "selection vector:\n",
    "\n",
    "1. pYPKa clones of terminator promoters cloned using the blunt ZraI restriction site\n",
    "2. pYPKa clones of genes to be expressed cloned using the blunt AjiI restriction site\n",
    "3. pYPKa clones of terminator promoters cloned using the blunt EcoRV restriction site\n",
    "4. pYPK0 acceptor vector\n",
    "\n",
    "### pCAPs\n",
    "\n",
    "pCAPs [Schlieper et al., 1998] is a small (3128 bp) positive selection vector for E. coli. It has the the usual ampicilllin resistance gene for selection and a collection of restriction sites in a gene encoding a toxic version of the cyclic AMP receptor protein, CRP. When DNA fragments are cloned within this gene, cell survival is permitted by the inactivation of the toxic CRP. Thus, there is positive selection for inserted fragments within the gene. This leads to very high frequency of correct clones in cloning experiments. pYPKa is a modified version carrying an additional AjiI restriction site.\n",
    "\n",
    "The pYPKa CRP gene has the following useful restriction enzyme sites in the order they appear in the pCAPs sequence file which is the opposite direction of the transcription of the CRP gene. The positions are given as distance between cut cites of the enzymes.\n",
    "\n",
    "\n",
    "| Enzyme | Pos |\n",
    "|--------|-----|\n",
    "| stop   | 0   |\n",
    "| XhoI   | 58  |\n",
    "| EcoRI  | 14  |\n",
    "| XbaI   | 19  |\n",
    "| MscI   | 21  |\n",
    "| SphI   | 33  |\n",
    "| NotI   | 60  |\n",
    "| ZraI   | 11  |\n",
    "| AjiI   | 50  |\n",
    "| EcoRV  | 32  |\n",
    "| AccIII |  2  |\n",
    "| MfeI   | 17  |\n",
    "| BsaAI  | 74  |\n",
    "| OliI   | 164 |\n",
    "| BamHI  | 70  |\n",
    "| start  | 28  |\n",
    "\n",
    "Restriction sites important for YPK are ZraI, AjiI, EcoRV and NotI\n",
    "\n",
    "    \n",
    "    \n",
    "    \n",
    "\n",
    "## Terminator-promoters\n",
    "A terminator promoter the piece of DNA separating two genes that are expressed in tandem. The first part of such an intergenic sequence contains the terminator of the upstream gene, and the latter part the promoter of the downstream gene.\n",
    "\n",
    "Figure 1. Tandemly expressed genes\n",
    "\n",
    "    ... Promoter> Gene 1 >Terminator Promoter> Gene 2 >Terminator ...\n",
    "\n",
    "## pYPK0\n",
    "The pYPK0 is a version of the pCAPs vector that has additional sequences for maintenance in Saccharomyces cerevisiae. The additional genes are URA3 for prototrophic selection for uracile and the 2µ sequence for replication.\n",
    "\n",
    "The pYPK0 was constructed from pCAPs and pSU0 as described in [Iizasa and Nagano 2006](http://www.ncbi.nlm.nih.gov/pubmed/16454044).\n",
    "\n",
    "\n",
    "Each single gene expression cassette (terminator-promoter-gene-terminator-promoter) can be released by digestion with NotI (present upstream of ZraI and PacI generated by the tp cloning strategyt (see below) cloning or by amplification with primers 775 and 778.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "## Strategy for cloning terminator promoters\n",
    "\n",
    "There are at least seven terminator-promoters (tps) that include strong promoters that have been reportedly used for expression of homologous or heterologous genes in S. cerevisiae (see table below). These tps range in size between 500 and 900 bp.\n",
    "\n",
    "| TP     | Expressed gene                             |\n",
    "|--------|--------------------------------------------|\n",
    "| ScENO2 | Enolase II                                 |\n",
    "| ScFBA1 | Fructose 1,6-bisphosphate aldolase         |\n",
    "| ScPDC1 | Pyruvate decarboxylase                     |\n",
    "| ScPGI1 | Phosphoglucose isomerase                   |\n",
    "| ScTDH3 | Glyceraldehyde-3-phosphate-dehydrogenase   |\n",
    "| ScTEF1 | Translational elongation factor EF-1 alpha |\n",
    "| ScTPI1 | Triose phosphate isomerase                 |\n",
    "\n",
    "Each tp is amplified with primers with six base pairs overhang. The overhangs\n",
    "are identical for all tps.\n",
    "\n",
    "\n",
    "Figure 2: Tails of primers used to amplify tps\n",
    "\n",
    "    >ScNNNNtpf\n",
    "    TTAAAT.....\n",
    "\n",
    "    >ScNNNNtpr_PacI\n",
    "    TAATTAA.....\n",
    "\n",
    "Each PCR product will have the following general structure (both strands shown):\n",
    "\n",
    "        TTAAAT- ... -TTAATTA\n",
    "        AATTTA- ... -AATTAAT\n",
    "\n",
    "\n",
    "The forward (tpf) primer adds a partial SmiI (ATTT^AAAT) site in front of the\n",
    "terminator of the tp PCR product. The reverse (tpr_PacI) primer adds a partial\n",
    "PacI (TTAAT^TAA) site at the end of the promoter. These sites are only active\n",
    "when cloned in the EcoRV site in pYPKa location as shown below:\n",
    "\n",
    "        ZraI                                              AjiI                           EcoRV\n",
    "    \n",
    "    -GAC                                                                                   \n",
    "        GTCgaggaacgccaggttgcccactttctcactagtgacctgcagccGAC                               \n",
    "                                                          GTGccatctgtgcagacaaacgcatcagGAT\n",
    "                                                                                         ATCcggattt-\n",
    "\n",
    "\n",
    "     ---- ZraI AjiI EcoRV -----\n",
    "    |                          |\n",
    "     ---pYPKa----ori amp-------\n",
    "\n",
    "pYPKa opened by ZraI or EcoRV:\n",
    "\n",
    "                ZraI                                EcoRV \n",
    "\n",
    "    ===GAC                   GTC===AjiI===GAT                   ATC=\n",
    "    \n",
    "    \n",
    "Terminator promoter sequences cloned in both locations:\n",
    "    \n",
    "\n",
    "    ===GAC TTAAAT-tp-TTAATTA GTC===AjiI===GAT TTAAAT-tp-TTAATTA ATC=\n",
    "                                           ---------    ---------\n",
    "                                           \n",
    "                                              SmiI         PacI\n",
    "                                              \n",
    "\n",
    "The entire single gene expression cassette (tp-gene-tp) can be cut out from the pYPK0 vector using [NotI](http://rebase.neb.com/rebase/enz/NotI.html) on the plasmid backbone and the [PacI](http://rebase.neb.com/rebase/enz/PacI.html) generated by the tp primer tails.\n",
    "\n",
    "The [SmiI](http://rebase.neb.com/rebase/enz/SmiI.html) site can be used to add for example c-terminal tags or as an alternative for pathway extension.\n",
    "\n",
    "### Gene cloning\n",
    "\n",
    "\n",
    "Tailed primers can be used to amplify any gene for recombination between any two\n",
    "terminator-promoters cloned in pYPKa.\n",
    "\n",
    "    >Gene_YPK_rec_fwd\n",
    "    gtcgaggaacgccaggttgcccactttctcactagtga-AA-ATG...\n",
    "                                              ---\n",
    "                                              start\n",
    "                                              codon\n",
    "                                              \n",
    "    <--------------  40 nt -------------->\n",
    "\n",
    "    >Gene_YPK_rec_rev\n",
    "    ATTTAAatcctgatgcgtttgtctgcacagatggCAC...\n",
    "                                         ---\n",
    "                                         stop\n",
    "                                         codon\n",
    "\n",
    "    <-------------- 38 nt--------------->\n",
    "\n",
    "Alternatively, a gene can be cloned in the blunt AjiI site of pYPKa by the same strategy as used for the tps.\n",
    "\n",
    "The gene can be released by ZraI + EcoRV digestion if no restriction sites for these genes are present in the gene, alternatively the gene with flanking sequences can be reamplified with primers that flank the cloned gene in the\n",
    "vector:\n",
    "\n",
    ">468_pCAPs_release_fw (25-mer)\n",
    "gtcgaggaacgccaggttgcccact\n",
    "\n",
    ">467_pCAPs_release_re (31-mer)\n",
    "ATTTAAatcctgatgcgtttgtctgcacaga\n",
    "\n",
    "The resulting cassette is similar to the one amplified with the tailed primers above, but the cloning primers can be made shorter and thereby less expensive and possibly more efficient.\n",
    "\n",
    "The following primers can be used to check presence and orientation of all\n",
    "terminator-promoters cloned in the ZraI or EcoRV site in pCAPs\n",
    "\n",
    ">775_tp_ZraI_fwd\n",
    "gcggccgctgacTTAAAT\n",
    ">776_tp_ZraI_rev\n",
    "cgttcctcgacTAATTAA\n",
    "\n",
    ">777_tp_Eco32I_fwd\n",
    "acgcatcaggatTTAAAT\n",
    ">778_tp_Eco32I_rev\n",
    "ggtaaatccggatTAATTAA\n",
    "\n",
    "                          \n",
    "\n",
    "\n",
    "\n",
    "How to make a TP-geneX-TP clone:\n",
    "\n",
    "Amplify the TP in a pCAPS_Z_tp with     167-567\n",
    "Amplify the TP in a pCAPS_A_geneX with  468-467\n",
    "Amplify the TP in a pCAPS_e_tp with     568-166\n",
    "\n",
    "co-transfrom the three PCR fragments with pCAPs-pSU0 \n",
    "digested w EcoRV + ZraI or amplified with nw1 and nw2.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "pYPK4   pCAPs-pSU0    HphMX4  2µ         ☹ (no AjiI) GR  ?????\n",
    "pYPK5   pYPKa-pSU0    HphMX4  2µ         ☹ AjiI      FP \n",
    "pYPK6   pCAPs-pSU0    HphMX4  2µ         ☹ (no AjiI) FP\n",
    "pYPK7   pCAPs-pSU0    kanMX4  2µ         ✔ (no AjiI) FP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
