{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Construction of the pYPKa vector\n",
    "The pYPKa is a version of the pCAPs vector where a unique [AjiI](http://rebase.neb.com/rebase/enz/AjiI.html) site has been created by site directed mutagenesis and a [AcsI](http://rebase.neb.com/rebase/enz/AcsI.html) site removed. The pCAPs plasmid was described by [Schlieper et al. 1998](http://www.ncbi.nlm.nih.gov/pubmed/9514792):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:32.913992Z",
     "iopub.status.busy": "2021-09-25T18:00:32.912610Z",
     "iopub.status.idle": "2021-09-25T18:00:32.933755Z",
     "shell.execute_reply": "2021-09-25T18:00:32.934187Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"80%\"\n",
       "            height=\"250\"\n",
       "            src=\"https://www.ncbi.nlm.nih.gov/pubmed/9514792\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "        ></iframe>\n",
       "        "
      ],
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x7f22d04f8670>"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# NBVAL_IGNORE_OUTPUT\n",
    "from IPython.display import IFrame\n",
    "IFrame('https://www.ncbi.nlm.nih.gov/pubmed/9514792', width='80%', height=250)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Importing the [pydna](https://pypi.python.org/pypi/pydna/) package.\n",
    "Pydna is [open source](https://github.com/BjornFJohansson/pydna), documentated [here](http://pydna.readthedocs.org/en/latest/) and has a support [forum](https://groups.google.com/forum/#!forum/pydna) as well as a publication:\n",
    "\n",
    "[Pereira F, Azevedo F, Carvalho Ã‚, Ribeiro GF, Budde MW, Johansson B: Pydna: a simulation and documentation tool for DNA assembly strategies using python. BMC Bioinformatics 2015, 16:142.](http://www.biomedcentral.com/1471-2105/16/142/abstract)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:32.940267Z",
     "iopub.status.busy": "2021-09-25T18:00:32.939336Z",
     "iopub.status.idle": "2021-09-25T18:00:33.692006Z",
     "shell.execute_reply": "2021-09-25T18:00:33.692384Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "from pydna.all import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The plasmid backbone sequence comes from a Genbank [record](http://www.ncbi.nlm.nih.gov/nuccore/AJ001614).\n",
    "Access to Genbank is needed in order to download the template.\n",
    "If you execute this script, change the email address below to your own.\n",
    "Always tell Genbank who you are, when using the service."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:33.703242Z",
     "iopub.status.busy": "2021-09-25T18:00:33.702540Z",
     "iopub.status.idle": "2021-09-25T18:00:33.705532Z",
     "shell.execute_reply": "2021-09-25T18:00:33.706180Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "gb = Genbank(\"bjornjobbb@gmail.com\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:33.710892Z",
     "iopub.status.busy": "2021-09-25T18:00:33.710357Z",
     "iopub.status.idle": "2021-09-25T18:00:33.716804Z",
     "shell.execute_reply": "2021-09-25T18:00:33.717608Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "pCAPs = gb.nucleotide(\"AJ001614.1\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:33.729474Z",
     "iopub.status.busy": "2021-09-25T18:00:33.728155Z",
     "iopub.status.idle": "2021-09-25T18:00:33.731506Z",
     "shell.execute_reply": "2021-09-25T18:00:33.730825Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "assert pCAPs.cseguid() == \"S6GoB_Z13VCfGDTEqNclu4It614\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We read the file from a local copy with more features defined."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:33.735670Z",
     "iopub.status.busy": "2021-09-25T18:00:33.734664Z",
     "iopub.status.idle": "2021-09-25T18:00:33.751411Z",
     "shell.execute_reply": "2021-09-25T18:00:33.751917Z"
    }
   },
   "outputs": [],
   "source": [
    "pCAPs = read(\"pCAPs.gb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:33.761279Z",
     "iopub.status.busy": "2021-09-25T18:00:33.759488Z",
     "iopub.status.idle": "2021-09-25T18:00:33.763304Z",
     "shell.execute_reply": "2021-09-25T18:00:33.762926Z"
    }
   },
   "outputs": [],
   "source": [
    "assert pCAPs.cseguid() == \"S6GoB_Z13VCfGDTEqNclu4It614\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The sequence to be mutated is contained withing the subsequence below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:33.767983Z",
     "iopub.status.busy": "2021-09-25T18:00:33.767120Z",
     "iopub.status.idle": "2021-09-25T18:00:33.772179Z",
     "shell.execute_reply": "2021-09-25T18:00:33.771672Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TTCTCACTAGTGACCTGCAGCCGGCGCGCCATCTGTGCAGACAAACGCA\n"
     ]
    }
   ],
   "source": [
    "print(str(pCAPs[555:604].seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:33.778301Z",
     "iopub.status.busy": "2021-09-25T18:00:33.777357Z",
     "iopub.status.idle": "2021-09-25T18:00:33.780378Z",
     "shell.execute_reply": "2021-09-25T18:00:33.780763Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CACGTC\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "from Bio.Restriction import AjiI\n",
    "print(AjiI.site)\n",
    "print(AjiI.is_palindromic())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Outline of the mutation strategy\n",
    "\n",
    "pCAPs sequence\n",
    "\n",
    "    TTCTCACTAGTGACCTGCAGCCGGCGCGCCATCTGTGCAGACAAACGCA   watson  \n",
    "    AAGAGTGATCACTGGACGTCGGCCGCGCGGTAGACACGTCTGTTTGCGT   crick\n",
    "                       <---ArgAla---\n",
    "\n",
    "Two aminoacids in the CRP coding sequence are indicated in the crick strand of pCAPs.\n",
    "\n",
    "\n",
    "Primers annealing to template\n",
    "                            \n",
    "    TTCTCACTAGTGACCTGCAGCCGGCGCGCCATCTGTGCAGACAAACGCA\n",
    "      |||||||||||||||||||||\n",
    "      gagtgatcactggacgtcggcTG\n",
    "                             GTGCcatctgtgcagacaaacg\n",
    "                                |||||||||||||||||||\n",
    "    AAGAGTGATCACTGGACGTCGGCCGCGCGGTAGACACGTCTGTTTGCGT\n",
    "    \n",
    "Resulting mutated seqence in pYPKa:    \n",
    "                                \n",
    "    ttctcactagtgacctgcagccgACGTGccatctgtgcagacaaacgca   watson\n",
    "    aagagtgatcactggacgtcggcTGCACggtagacacgtctgtttgcgt   crick\n",
    "                          ------                        AjiI site\n",
    "    \n",
    "                       <---ArgAla---\n",
    "    \n",
    "    \n",
    "The mutation is silent, the codons formed in the CRP orf on the crick strand encode the same aminoacids\n",
    "    \n",
    "          Ala Arg\n",
    "    \n",
    "    pCAPS gcg cgg\n",
    "    \n",
    "    pYPKa gca cgt\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:33.785801Z",
     "iopub.status.busy": "2021-09-25T18:00:33.783867Z",
     "iopub.status.idle": "2021-09-25T18:00:33.788441Z",
     "shell.execute_reply": "2021-09-25T18:00:33.789147Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "primers =parse('''\n",
    ">568_pCAPsAjiIR (22-mer)\n",
    "GTGCcatctgtgcagacaaacg\n",
    ">567_pCAPsAjiIF (23-mer)\n",
    "GTcggctgcaggtcactagtgag''', ds=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:33.804819Z",
     "iopub.status.busy": "2021-09-25T18:00:33.796905Z",
     "iopub.status.idle": "2021-09-25T18:00:34.056463Z",
     "shell.execute_reply": "2021-09-25T18:00:34.056842Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "pYPKa =pcr(primers, pCAPs).looped().synced(pCAPs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:34.063788Z",
     "iopub.status.busy": "2021-09-25T18:00:34.063075Z",
     "iopub.status.idle": "2021-09-25T18:00:34.065733Z",
     "shell.execute_reply": "2021-09-25T18:00:34.065241Z"
    }
   },
   "outputs": [],
   "source": [
    "assert pYPKa.cseguid() == \"aV1eIrzOiCjvw01yvKkxDXHKLMk\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:34.077986Z",
     "iopub.status.busy": "2021-09-25T18:00:34.077329Z",
     "iopub.status.idle": "2021-09-25T18:00:34.081005Z",
     "shell.execute_reply": "2021-09-25T18:00:34.080497Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "cSEGUID_aV1eIrzOiCjvw01yvKkxDXHKLMk"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pYPKa.stamp()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The first of the pCAPs features is the crp gene. This is left out in the pYPKa due to the PCR process. \n",
    "It is added back to the pYPKa features below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:34.086089Z",
     "iopub.status.busy": "2021-09-25T18:00:34.084832Z",
     "iopub.status.idle": "2021-09-25T18:00:34.087518Z",
     "shell.execute_reply": "2021-09-25T18:00:34.088295Z"
    }
   },
   "outputs": [],
   "source": [
    "crp_feature = pCAPs.features[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:34.095292Z",
     "iopub.status.busy": "2021-09-25T18:00:34.094348Z",
     "iopub.status.idle": "2021-09-25T18:00:34.099055Z",
     "shell.execute_reply": "2021-09-25T18:00:34.098441Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['artificially mutated crp; cyclic AMP receptor protein']"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "crp_feature.qualifiers[\"note\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:34.104589Z",
     "iopub.status.busy": "2021-09-25T18:00:34.103918Z",
     "iopub.status.idle": "2021-09-25T18:00:34.106260Z",
     "shell.execute_reply": "2021-09-25T18:00:34.106690Z"
    }
   },
   "outputs": [],
   "source": [
    "pYPKa.features.append(crp_feature)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:34.110660Z",
     "iopub.status.busy": "2021-09-25T18:00:34.110054Z",
     "iopub.status.idle": "2021-09-25T18:00:34.113537Z",
     "shell.execute_reply": "2021-09-25T18:00:34.112984Z"
    }
   },
   "outputs": [],
   "source": [
    "from Bio.Restriction import EcoRV, ZraI, AjiI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:34.117651Z",
     "iopub.status.busy": "2021-09-25T18:00:34.116979Z",
     "iopub.status.idle": "2021-09-25T18:00:34.120588Z",
     "shell.execute_reply": "2021-09-25T18:00:34.119927Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [],
   "source": [
    "pYPKa.name = \"pYPKa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false,
    "execution": {
     "iopub.execute_input": "2021-09-25T18:00:34.127004Z",
     "iopub.status.busy": "2021-09-25T18:00:34.125761Z",
     "iopub.status.idle": "2021-09-25T18:00:34.141283Z",
     "shell.execute_reply": "2021-09-25T18:00:34.141803Z"
    },
    "jupyter": {
     "outputs_hidden": false
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<font face=monospace><a href='pYPKa.gb' target='_blank'>pYPKa.gb</a></font><br>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pYPKa.write(\"pYPKa.gb\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:bjorn37]",
   "language": "python",
   "name": "conda-env-bjorn37-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
